### PR TITLE
[Model][PP] Validate DeepSeek-V4.1 sharing dependencies before construction

### PR DESCRIPTION
**Pull request by 0z5a**
## Purpose

Resolve each KV, index-key, top-k and candidate dependency before constructing pipeline layers. Invalid source lists and stage cuts now report the source/consumer layers and stages instead of failing later during layer/cache setup. Valid sharing-group-aligned cuts continue to work.

Also allocate cache tensors only for a projected group's local `layer_names`: an empty `UniformTypeKVCacheSpecs` group must not allocate its remote layers. This allocation repair is a prerequisite shared with #56222 and #56223. This PR does not enable cross-stage transfer; #56223 contains that separate opt-in implementation.

Searched open V4.1/pipeline-sharing PRs and the #56214 references. No other open PR found implements this V4.1 source-dependency precheck plus empty projected-group allocation fix.

Based on #56214 (`dsv41-feat`, `c9d909e802a39292f54101bff8a36096761ea605`). The PR targets that feature branch so the model implementation is not repeated in this diff.

Rebased on 2026-09-11 after the feature branch was rewritten. Changed-file Python parsing and all applicable pre-commit hooks passed. The pipeline dependency suite passed (11 CPU tests). GPU/full-model measurements below belong to the prior revision based on `e47aa780bccf59f59dfa2cbb18e17a10b4fe69ba`; they were not rerun on this rebased head.

AI assistance: OpenAI Codex assisted with implementation, review, test execution and preparation of this PR. This draft does not claim that a human has completed a line-by-line review.

## Test Plan

Fresh split: **11 CPU tests passed** in `tests/models/test_deepseek_v41_pipeline.py`, using a standalone dependency-light loader (`pytest -q --noconftest -o addopts=` after loading the actual pure dependency module). All applicable pre-commit hooks and changed-file Python parsing passed.

The integration also ran the projected-cache-group tests in `tests/v1/core/test_kv_cache_utils.py` on H100. The allocation implementation and its regression test are unchanged by the split.

## Test Result

Recorded full-model runs used `deepseek-ai/DeepSeek-V4.1-Flash`, revision `df42c109f1defefcbfcedbe7d905718a12266e40`, on 4×SXM H100 80 GiB. Each positive matrix case checked six short questions, cold/cached long input with chunked prefill, and four unequal-length concurrent requests. “Text equal” means short-answer text equality to the TP4 reference, not logits/token-probability parity or a standard accuracy benchmark.

**Evidence scope:** GPU results below come from the earlier combined integration based on #56214, including companion changes. The isolated PR has not had a new full-model GPU run. Fresh split checks are listed separately; passing the integrated run does not establish isolated-branch equivalence.

| Integrated configuration / check | Observed result | Meaning |
| --- | --- | --- |
| PP2×TP2, 20/20 layers | QA 6/6 correct and text-equal; long/prefix 2/2; concurrent 4/4 | Valid partition completes full-model serving checks |
| PP4×TP1, equal 10/10/10/10 layers, sharing disabled | Expected dependency rejection before layer construction | Invalid partition fails early with source/consumer information |
| H100 projected-cache-group regression tests | 2/2 PASS | Empty stage-local groups do not allocate remote-layer caches |

| Fresh split validation | Result |
| --- | --- |
| Pure pipeline dependency tests | 11/11 CPU PASS, using the standalone loader below |
| Applicable pre-commit hooks and changed Python parsing | PASS; 5 changed files |

This change establishes partition validation and correct local allocation. No measured throughput or GPU-memory reduction is attributed to it.

---
<details>
<summary>PR description checklist</summary>

- [x] Purpose and related work described.
- [x] Test plan and actual results stated.
- [x] Model evaluation limitations stated.
- [x] AI assistance disclosed.

</details>

<details>
<summary>Standalone CPU validation loader</summary>

Saved as `run_cpu_checks.py`; set `DSV41_SPLIT_REPO` to the checked-out branch and pass the test paths listed above. This avoids CUDA model-registry imports on macOS while executing the actual leaf-module implementation, tests and tensor/collective operations. It is not a CUDA or full-model test.

```python
"""Run unchanged CPU test functions against split source without CUDA imports.

Preload only dependency-light leaf modules, as in the integration validation
driver. CUDA-only tests keep their own skip conditions. No tensor operations or
collectives are mocked by this loader.
"""
import importlib.util
import os
from pathlib import Path
import sys

import pytest

ROOT = Path(os.environ["DSV41_SPLIT_REPO"])
for name in (
    "vllm.models.deepseek_v4_1.common.pipeline",
    "vllm.models.deepseek_v4_1.common.pipeline_transfer",
    "vllm.v1.worker.ubatch_inputs",
):
    path = ROOT / (name.replace(".", "/") + ".py")
    if not path.is_file():
        continue
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    sys.modules[name] = module
    spec.loader.exec_module(module)

if __name__ == "__main__":
    raise SystemExit(pytest.main(["-q", "--noconftest", "-o", "addopts=", *sys.argv[1:]]))
```

</details>

### CODE REVIEW

**Comment by 0z5a** [comment]
Validation follow-up for `4f8b7e3872b878e76bffff2b46a03173961f89a5`: the 11 current-head CPU dependency tests support the validation logic. The documented GPU/model results belong to an earlier combined integration, and the recent H200 attempt did not add an isolated full-model run on this head.

This is a partition-validation/local-allocation fix, so a throughput increase is not its acceptance criterion. The useful missing evidence is a current-head legal PP2×TP2 20/20 run with long/chunked and cached requests, an invalid PP4 10/10/10/10 partition rejected before construction with the expected dependency details, and the projected-cache regression proving that empty stage-local groups allocate no remote-layer tensors. GPU allocation receipts would support that specific allocation claim; different PP topologies should not be presented as an isolated speedup.

Contributions from anyone with access to sufficient GPU capacity are very welcome: reproducible GPU E2E benchmarks, targeted correctness checks, and focused fixes. Please include source/model revisions, GPU topology, offload/KV settings, exact commands, and raw results; repeat timing runs when making performance claims. I am happy to review and cherry-pick fixes while preserving the original authorship and clear attribution.

**Comment by 0z5a** [comment]
Resubmitted as #56437 against `main`, since `dsv41-feat` was merged into `main` in #56214.
